Fix calculation of missing balance in API error

Anviking opened the pull request:
# Issue Number

ADP-697

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] Add new `difference` function for `TokenMap` and `TokenBundle`
- [x] Calculate the missing balance using `required - available` using `difference` rather than `subtract` 
- [x] TODO: More tests?
- [ ] TODO: Can we property test the API handler?
    - Since this logic is in the API rendering of the errors, it seems very tricky.


# Comments

Analysis of problem:

In Server.hs there's this calculation:

```
                        missing
                            = TokenBundle.Flat
                            $ fromMaybe TokenBundle.empty
                            $ TokenBundle.subtract
                                balanceRequired balanceAvailable
```

But because of this requirement of TokenBundle.subtract
```haskell
-- Returns 'Nothing' if the second bundle is not less than or equal to the first
-- bundle when compared with the `leq` function.
```
it should mean that if we own tokens that are not required, then we hit the nothing case, and we get a missing `coin: 0 tokens: []`.

Think we need something like this:
```haskell
difference :: TokenMap -> TokenMap -> TokenMap
difference = modifySimple $ Map.differenceWith TokenQuantity.subtract
  where
    modifySimple
        :: (Map AssetId TokenQuantity -> Map AssetId TokenQuantity -> Map AssetId TokenQuantity)
        -> TokenMap
        -> TokenMap
        -> TokenMap
    modifySimple f x y = fromFlatList . Map.toList
        $ f (Map.fromList . toFlatList $ x)
            (Map.fromList . toFlatList $ y)
```

<!-- Additional comments or screenshots to attach if any -->

<!--
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Jira will detect and link to this PR once created, but you can also link this PR in the description of the corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
 ✓ Finally, in the PR description delete any empty sections and all text commented in <!--, so that this text does not appear in merge commit messages.
-->

Anviking committed:
Add {TokenMap,TokenBundle}.difference functions

Improve difference impl and tests

Further tweaks to tests

Anviking committed:
Add balanceMissing function

And test it in performSelection properties.

Anviking committed:
Use new balanceMissing when rendering BalanceInsufficientErrors

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/Types/TokenBundle.hs:
```diff
@@ -293,6 +294,13 @@ add (TokenBundle (Coin c1) m1) (TokenBundle (Coin c2) m2) =
 subtract :: TokenBundle -> TokenBundle -> Maybe TokenBundle
 subtract a b = guard (b `leq` a) $> unsafeSubtract a b
 
+difference :: TokenBundle -> TokenBundle -> TokenBundle
```
Should describe the difference compared to `subtract` — which I'm, to be honest, not entirely sure of.

Anviking replied:
Updated. Although, not with emphasis on the difference to `subtract`.

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/Types/TokenBundle.hs:
```diff
@@ -293,6 +294,13 @@ add (TokenBundle (Coin c1) m1) (TokenBundle (Coin c2) m2) =
+difference :: TokenBundle -> TokenBundle -> TokenBundle
+difference (TokenBundle (Coin c1) m1) (TokenBundle (Coin c2) m2) =
+    TokenBundle
+        (Coin $ c1 - c2)
```
Should add the properties for TokenMap to TokenBundle too. This `-` seems dangerous.

Anviking replied:
Fixed.

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/Types/TokenMapSpec.hs:
```diff
@@ -255,6 +261,19 @@ prop_subtract_invariant m1 m2 = property $
+
+prop_difference_leq :: TokenMap -> TokenMap -> Property
+prop_difference_leq x y = property $
+    x `difference` y `leq` x
```
Should the reverse not also be true?

I.e.:

```haskell
x `difference` y `leq` y
```

And I'm guessing there could be a symmetric property:

```hs
x `difference` y = y `difference` x
```

Anviking replied:
If `x = {a -> 1}` and `y = {}`, then `x - y = {a -> 1}`, which is larger than `{}`

```haskell
       Falsifiable (after 8 tests and 8 shrinks):
         TokenMap (fromList [(UnsafeTokenPolicyId (Hash "\170\170\170\170\170\170\170\170\170\170\170\170\170\170\170\170\170\170\170\170\170\170\170\170\170\170\170\170"),NonEmptyMap {least = (UnsafeTokenName "TokenA",TokenQuantity 1), rest = fromList []})])
         TokenMap (fromList [])
```

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/Types/TokenMapSpec.hs:
```diff
@@ -255,6 +261,19 @@ prop_subtract_invariant m1 m2 = property $
+prop_difference_add :: TokenMap -> TokenMap -> Property
+prop_difference_add x y = property $
+    (x `difference` y) `TokenMap.add` y
+        === x
```
I think in general this isn't true, unless this is also true:

```hs
y `leq` x
```

Anviking replied:
Yup, I just noticed.

jonathanknowles replied:
The definition of `geq` should also handle the equality case:

```hs
geq a b = a == b || (not $ leq a b)
```

Anviking replied:
I went with
```haskell
b `leq` a
```

Anviking committed:
Tweak formatting

Anviking commented on lib/core/test/unit/Cardano/Wallet/Primitive/Types/TokenBundleSpec.hs:
```diff
@@ -39,6 +41,46 @@ spec =
+    describe "Arithmetic" $ do
+        it "prop_difference_zero (x - 0 = x)" $
+            property prop_difference_zero
+        it "prop_difference_zero2 (0 - x = 0)" $
```
Didn't want to break the naming convention too much, but I _did_ want to provide more readable expressions.

Anviking replied:
Also, prop_difference_zero{1,2,3} could be unified into a single property, but I think this might be simpler/more readable in some sense.

jonathanknowles replied:
I think it's fine to have separate properties here. 👍🏻

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Types/TokenBundle.hs:
```diff
@@ -293,6 +297,31 @@ add (TokenBundle (Coin c1) m1) (TokenBundle (Coin c2) m2) =
 subtract :: TokenBundle -> TokenBundle -> Maybe TokenBundle
 subtract a b = guard (b `leq` a) $> unsafeSubtract a b
 
+
```
```suggestion
```

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Types/TokenBundle.hs:
```diff
@@ -293,6 +297,31 @@ add (TokenBundle (Coin c1) m1) (TokenBundle (Coin c2) m2) =
+        (fromMaybe (Coin 0) $ Coin.subtractCoin c1 c2)
+        (TokenMap.difference m1 m2)
+
+
```
```suggestion

```

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Types/TokenBundle.hs:
```diff
@@ -293,6 +297,31 @@ add (TokenBundle (Coin c1) m1) (TokenBundle (Coin c2) m2) =
+difference (TokenBundle c1 m1) (TokenBundle c2 m2) =
+    TokenBundle
+        (fromMaybe (Coin 0) $ Coin.subtractCoin c1 c2)
+        (TokenMap.difference m1 m2)
```
👍🏻

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Types/TokenMap.hs:
```diff
@@ -512,6 +513,39 @@ add a b = F.foldl' acc a $ toFlatList b
+        -> TokenMap
+        -> TokenMap
+        -> TokenMap
+    modifySimple f x y = fromFlatMap (f (toFlatMap x) (toFlatMap y))
```
This function can actually be implemented in terms of `adjustQuantity`:

```suggestion
difference m1 m2 = L.foldl' reduce m1 (toFlatList m2)
  where
    reduce :: TokenMap -> (AssetId, TokenQuantity) -> TokenMap
    reduce m (a, q) = adjustQuantity m a
        (fromMaybe TokenQuantity.zero . (`TokenQuantity.subtract` q))
```

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobinSpec.hs:
```diff
@@ -593,6 +600,18 @@ prop_performSelection_small minCoinValueFor costFor (Blind (Small criteria)) =
+    assetsInUTxO = not
+        . Set.null
+        . UTxOIndex.assets
+        $ utxoAvailable criteria
```
Suggestion: make this name a bit more explicit.

Justification: with the name `assetsInUTxO`, it might seem to the reader that this is the set of assets included in the UTxO.

```suggestion
    utxoHasAtLeastOneAsset = not
        . Set.null
        . UTxOIndex.assets
        $ utxoAvailable criteria
```

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobinSpec.hs:
```diff
@@ -593,6 +600,18 @@ prop_performSelection_small minCoinValueFor costFor (Blind (Small criteria)) =
+        . UTxOIndex.assets
+        $ utxoAvailable criteria
+
+    assetsInOutputs = not . Set.null $ TokenBundle.getAssets outputTokens
```
Suggestion: make this name a bit more explicit.

Justification: with the name `assetsInOutputs`, it might seem to the reader that this is the set of assets included in the outputs.

```suggestion
    outputsHaveAtLeastOneAsset =
        not . Set.null $ TokenBundle.getAssets outputTokens
```

jonathanknowles reviewed (commented):
Hi @Anviking 

Thanks for doing this!

I think this looks good so far. I have a few suggestions which I've left in comments.

I think there are a couple of things missing:

1. a property that tests **_what the result should be_**. In other words, we should test that the quantities produced by `difference a b` are equivalent to performing `TokenQuantity.subtract` on each pair of token quantities, with a default value of `0` for values that would be negative.
2. a property that checks that `difference` preserves the invariant. We currently do this for all of the public operations exported by `TokenMap`. Something like:
    ```hs
    prop_difference_invariant :: TokenMap -> TokenMap -> Property
    prop_difference_invariant m1 m2 =
        property $ invariantHolds $ TokenMap.difference m1 m2
    ```

Lastly, the properties for `difference` should probably be in their own section. (Currently, for `TokenMap`, they're lumped in with the `Invariant properties` section.)

Anviking committed:
Apply suggestions from code review

Co-authored-by: Jonathan Knowles <jonathan.knowles@iohk.io>

Anviking committed:
Add prop_difference_invariant

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -2873,23 +2873,11 @@ instance LiftHandler ErrSelectAssets where
-                            ]
+                    apiError err403 NotEnoughMoney $ mconcat
+                        [ "I can't process this payment as there are not "
+                        , "enough funds available in the wallet. I am only "
```
Suggestion: reword this slightly.

Justification: it's possible the user is missing a large amount!

```suggestion
                        , "enough funds available in the wallet. I am "
```

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/Types/TokenMapSpec.hs:
```diff
@@ -255,6 +268,37 @@ prop_subtract_invariant m1 m2 = property $
 
+prop_difference_zero :: TokenMap -> Property
+prop_difference_zero x =
+    x `difference` mempty === x
```
This property (and others) are currently in the "Invariant properties" section (starting at line 228).

They should probably be moved to the "Arithmetic properties" section.

See: https://github.com/input-output-hk/cardano-wallet/pull/2548/commits/a9c2f9a578b930e8c6100a386f09cbbadefc037d

Anviking replied:
Ah! I was looking at the HSpec tree — thanks!

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/Types/TokenMapSpec.hs:
```diff
@@ -255,6 +268,37 @@ prop_subtract_invariant m1 m2 = property $
+
+prop_difference_invariant :: TokenMap -> TokenMap -> Property
+prop_difference_invariant m1 m2 =
+    property $ invariantHolds $ TokenMap.difference m1 m2
```
This property actually **_is_** an invariant property, so it can remain where it is!

jonathanknowles committed:
Move `difference` properties into the "Arithmetic properties" section.

These properties were in the "Invariant properties" section, which
should only contain properties relating to testing the invariant.

jonathanknowles committed:
Add property `prop_difference_subtract`.

This test demonstrates that `difference a b` is always equivalent to
`subtract a b` in the case where `leq b a`.

jonathanknowles committed:
Add property `prop_difference_equality`.

jonathanknowles reviewed (commented):
Hi @Anviking 

Thanks for updating your PR!

In another PR (#2548) based on this one, I've added a couple of properties for the `difference` function (and added you as a reviewer).

Anviking committed:
Re-word API error message

Co-authored-by: Jonathan Knowles <jonathan.knowles@iohk.io>

jonathanknowles reviewed (approved):
Hi @Anviking

LGTM!

Thanks again for making further changes.
